# Project 2 - Part 1

## Introduction

Part 1 of this project, itself, has two main aspects. The first is to load and filter the business.json file that contains the details of businesses and their review scores. The second is to obtain, and clean, data of the member states of the USA along with their populations. These two parts are interlinked as you will be using the latter to guide you in filtering the former.

Main skills:
* getting and loading data
* cleaning/wrangling data
* filtering rows and columns
* saving the preprocessed data

## Business reviews

The yelp challenge data can be downloaded from the website [here](https://www.yelp.com/dataset/challenge). Extracting the files, you should have a `business.json` file.

* How many countries does the yelp website say are in the data?

The website says there are 2 countries.

### Instructions

* Read business.json into a pandas DataFrame
* How big is the resultant DataFrame? How many rows and columns?
* Print the first few rows to get a sense of the contents

In [1]:
import pandas as pd

In [2]:
business = pd.read_json('./data/business.json', lines=True)

In [3]:
business.shape

(192609, 14)

In [4]:
business.head()

address  \
0     2818 E Camino Acequia Drive   
1            30 Eglinton Avenue W   
2       10110 Johnston Rd, Ste 15   
3   15655 W Roosevelt St, Ste 237   
4  4209 Stuart Andrew Blvd, Ste F   

                                          attributes             business_id  \
0                           {'GoodForKids': 'False'}  1SWheh84yJXfytovILXOAQ   
1  {'RestaurantsReservations': 'True', 'GoodForMe...  QXAEGFB4oINsVuTFxEYKFQ   
2  {'GoodForKids': 'True', 'NoiseLevel': 'u'avera...  gnKjwL_1w79qoiV3IC_xQQ   
3                                               None  xvX2CttrVhyG2z1dFg_0xw   
4  {'BusinessAcceptsBitcoin': 'False', 'ByAppoint...  HhyxOkGAM07SRYtlQ4wMFQ   

                                          categories         city  \
0                                  Golf, Active Life      Phoenix   
1  Specialty Food, Restaurants, Dim Sum, Imported...  Mississauga   
2                  Sushi Bars, Restaurants, Japanese    Charlotte   
3                      Insurance, Financial Services     Goodyear   
4  Plumbing, Shopping, Local Services, Home Servi...    Charlotte   

                                               hours  is_open   latitude  \
0                                               None        0  33.522143   
1  {'Monday': '9:0-0:0', 'Tuesday': '9:0-0:0', 'W...        1  43.605499   
2  {'Monday': '17:30-21:30', 'Wednesday': '17:30-...        1  35.092564   
3  {'Monday': '8:0-17:0', 'Tuesday': '8:0-17:0', ...        1  33.455613   
4  {'Monday': '7:0-23:0', 'Tuesday': '7:0-23:0', ...        1  35.190012   

    longitude                             name postal_code  review_count  \
0 -112.018481       Arizona Biltmore Golf Club       85016             5   
1  -79.652289       Emerald Chinese Restaurant     L5R 3E7           128   
2  -80.859132      Musashi Japanese Restaurant       28210           170   
3 -112.395596  Farmers Insurance - Paul Lorenz       85338             3   
4  -80.887223              Queen City Plumbing       28217             4   

   stars state  
0    3.0    AZ  
1    2.5    ON  
2    4.0    NC  
3    5.0    AZ  
4    4.0    NC

What distinct states do we have, and how many reviewed businesses do we have in each?

In [5]:
states_counts = business.state.value_counts(ascending=False)
states_counts = states_counts.reset_index(name='business_count').rename(columns={'index': 'state'})
states_counts

state  business_count
0     AZ           56686
1     NV           36312
2     ON           33412
3     NC           14720
4     OH           14697
5     PA           11216
6     QC            9219
7     AB            8012
8     WI            5154
9     IL            1932
10    SC            1162
11    NY              22
12    CA              19
13    TX               6
14   XGM               4
15    FL               4
16    AL               3
17    WA               3
18    CT               3
19    NE               2
20    GA               2
21    VT               2
22   XWY               2
23    AK               2
24    VA               2
25   DUR               1
26   XGL               1
27    NJ               1
28   BAS               1
29    AR               1
30    UT               1
31    BC               1
32    TN               1
33   CON               1
34   DOW               1
35    NM               1

As a non-American, even I know that XGM is not an abbreviation for any state in the USA. What are those three letter states? There aren't many of them so let's print out the address, city, latitude, longitude, postal_code, and state for them.

In [6]:
len(business.state)

192609

In [7]:
business.loc[business.state.apply(len) == 3, ['address', 'city', 'latitude', 'longitude', 'postal_code', 'state']]

address         city   latitude  \
27744                                    8 York St         Bath  43.640646   
33265   The Old Sorting Office, 12 Hayfield Street         Sale  42.996059   
35255                                                      Bury  42.996059   
39022               Gorebrook Works, Pinkbank Lane   Manchester  42.996059   
54920                          37 Monk Bridge Road        Leeds  45.456999   
58669                   The Cove, Coverack Helston  Church Cove  35.532021   
127476               1136 Center Street, Suite 442    Thornhill  43.808563   
133170                             61 Queen Street        Leeds  43.652821   
144601                                                   Oldham  42.996059   
179148                   110 Islington High Street       London  43.645355   
188285                Odyssey Arena, 2 Queens Quay         Down  43.641658   

        longitude postal_code state  
27744  -79.380939     BA1 1NG   BAS  
33265  -89.568889     M33 7XW   XGM  
35255  -89.568889     BL8 4DR   XGM  
39022  -89.568889     M12 5GH   XGM  
54920  -73.595250     LS6 4EP   XWY  
58669  -80.851682    TR12 6SX   CON  
127476 -79.463806     L4J 3M8   DUR  
133170 -79.376345    LS27 8EB   XWY  
144601 -89.568889     OL2 6PX   XGM  
179148 -79.524467      N1 8EG   XGL  
188285 -79.376030      BT33 0   DOW

Okay, clearly these are in the UK. We won't be needing those. But what of the two-letter states? Are they all valid for USA? What we really need is a list of the abbreviations of the states. We also think that the populations of the states will be useful. Let's find that data next.

## Abbreviations and populations of USA member states

The details we want can be found in a table in [this page](https://simple.wikipedia.org/wiki/List_of_U.S._states) on wikipedia.

* Read in the table
* Inspect it - what do you start with?
* Produce a DataFrame with two columns: 'state' (abbreviation) and 'population'
* Check you have the expected number of rows (states)

In [8]:
usa_states = pd.read_html('https://simple.wikipedia.org/wiki/List_of_U.S._states')

In [9]:
type(usa_states)

list

In [10]:
len(usa_states)

1

In [11]:
usa_states = usa_states[0]
usa_states.head()

0           1           2             3   \
0   Name &        Abbr      Cities   Established   
1  Capital  Largest[3]         mi2           km2   
2  Alabama          AL  Montgomery    Birmingham   
3   Alaska          AK      Juneau     Anchorage   
4  Arizona          AZ     Phoenix  Feb 14, 1912   

                             4              5             6              7   \
0  Population[upper-alpha 1][1]  Total area[2]  Land area[2]  Water area[2]   
1                           mi2            km2           mi2            km2   
2                  Dec 14, 1819        4874747         52420         135767   
3                   Jan 3, 1959         739795        665384        1723337   
4                       7016270         113990        295234         113594   

               8          9        10        11   12  
0  Numberof Reps.        NaN      NaN       NaN  NaN  
1             NaN        NaN      NaN       NaN  NaN  
2           50645   131171.0   1775.0    4597.0  7.0  
3          570641  1477953.0  94743.0  245384.0  1.0  
4          294207      396.0   1026.0       9.0  NaN

Okay, this is rather messy because of the formatting of the original table. We have some cleaning to do. The headers (column names) are spread over two rows and the population count can be in one of two columns, depending on what happened to the left. Patterns that helps us here is that the population count is the first numeric-only column and it's in column 4 (at the earliest, if the capital city is also the largest city) or column 5.

In [12]:
# define a convenient function that plucks out the population count from each row
def extract_population(row):
    import re
    return(row[4] if re.match('\d+', row[4]) else row[5])

In [13]:
# get rid of the first two rows and we aren't concerned with anything after the fifth column
usa_states_popn = usa_states.iloc[2:, :6]
usa_states_popn.iloc[:, 4] = usa_states_popn.apply(lambda x: extract_population(x), axis=1)
usa_states_popn = usa_states_popn.iloc[:, [1, 4]].reset_index(drop=True)
usa_states_popn.columns = ['state', 'population']
usa_states_popn['population'] = pd.to_numeric(usa_states_popn['population'])

In [14]:
usa_states_popn.shape

(50, 2)

In [15]:
usa_states_popn.head()

state  population
0    AL     4874747
1    AK      739795
2    AZ     7016270
3    AR     3004279
4    CA    39536653

## Combining the data - what states do we have and not have?

In [16]:
business_states = business['state'].drop_duplicates().reset_index(drop=True)
business_states = set(business_states)
print(business_states)

{'XWY', 'XGL', 'OH', 'WA', 'NM', 'AR', 'ON', 'BC', 'TN', 'AK', 'NJ', 'VT', 'SC', 'CA', 'TX', 'DOW', 'FL', 'CT', 'XGM', 'AL', 'AB', 'AZ', 'NE', 'GA', 'PA', 'WI', 'NC', 'IL', 'CON', 'QC', 'BAS', 'NY', 'VA', 'NV', 'UT', 'DUR'}


In [17]:
all_usa_states = set(usa_states_popn.state)

### States we have

In [18]:
states_we_have = all_usa_states & business_states
n_states_we_have = len(states_we_have)

In [19]:
print('We have the '+str(n_states_we_have)+' states:')
print(states_we_have)

We have the 25 states:
{'OH', 'WA', 'NM', 'AR', 'TN', 'AK', 'NJ', 'VT', 'SC', 'CA', 'TX', 'CT', 'FL', 'AL', 'AZ', 'NE', 'GA', 'PA', 'WI', 'NC', 'IL', 'NY', 'VA', 'NV', 'UT'}


### States we don't have

In [20]:
states_not_have = all_usa_states - business_states
n_states_not_have = len(states_not_have)

In [21]:
print('We do not have the '+str(n_states_not_have)+' states:')
print(states_not_have)

We do not have the 25 states:
{'IN', 'KS', 'LA', 'DE', 'MT', 'OR', 'IA', 'MO', 'CO', 'HI', 'MS', 'KY', 'MI', 'ND', 'ID', 'ME', 'WY', 'MN', 'SD', 'MA', 'MD', 'OK', 'RI', 'NH', 'WV'}


### States we have that we don't want (non-USA)

In [22]:
extra_states = business_states - all_usa_states
n_extra_states = len(extra_states)

In [23]:
print('Extra states in the business data are the ' + str(n_extra_states) + ' states:')
print(extra_states)

Extra states in the business data are the 11 states:
{'BAS', 'XGM', 'AB', 'XWY', 'DUR', 'XGL', 'QC', 'CON', 'DOW', 'ON', 'BC'}


* Do you recognise these other two-character states?
* What does this tell you about the information on the yelp website?

QC is Quebec, AB is Alberta, BC is British Colombia, and ON is Ontario. So we have Canada in here as well as USA and those few from UK. This tells us the statement on the website is not totally accurate. You should always validate what you are told about data!

## Take stock of the data we have

* Think about the picture we have of our data
    * What does it contain?
    * What is it lacking?
    * Does this make us wary about using it?

We know we only have 25 of the 50 states. We have also seen that of the 25 we have, the number of reviews is very unequal and many of those states only have 1 or 2 reviews. Therefore we probably don't have a random sample. This should make us wary about drawing inferences about the population. Even for the states we seem to have quite a bit of review data on, those reviews may not have been randomly drawn from those states!

## Filter and keep just USA states with at least 1000 reviewed businesses

Recall we previously calculated the number of reviewed businesses per state. We'll filter at the level of 1000 businesses per state as below that count we dropped to only a handful of businesses.

In [24]:
reviewed_states = set(states_counts.loc[states_counts.business_count > 1000, 'state'])

In [25]:
all_usa_states & reviewed_states

{'AZ', 'IL', 'NC', 'NV', 'OH', 'PA', 'SC', 'WI'}

In [26]:
usa_business = business.loc[business.state.isin(all_usa_states & reviewed_states), :]
usa_business.shape

(141879, 14)

So we've filtered our 192609 rows down to just the 141879 associated with the states in USA where there were at least 1000 reviewed businesses.

## Business categories

We will want to know what category our businesses are. How many of our businesses in the USA in this dataset do not have any category associated with them?

In [27]:
sum(usa_business.categories.isnull())

368

* display the first few entries with no category

In [28]:
usa_business.loc[usa_business.categories.isnull(), :].head()

address attributes             business_id categories  \
265          6340 E Thomas Rd       None  xOR85RicYj642O3_iJ7hgg       None   
1020         995 Greentree Rd       None  llG2YxnevtFhW7LVa4QJXQ       None   
1332  1816 W Pointe Dr, Ste C       None  zi2CV7isWyGRxVXJxdDQXg       None   
2399           1724 W 10th Pl       None  AEpvBhxrmHJud4UgvF2yWw       None   
2909             3434 Kier Rd       None  AMhPJSh9S6Qz9Q3DuwPgtA       None   

                 city hours  is_open   latitude   longitude  \
265        Scottsdale  None        1  33.480373 -111.945650   
1020       Pittsburgh  None        1  40.417374  -80.044817   
1332        Charlotte  None        1  35.257508  -80.972256   
2399            Tempe  None        1  33.418006 -111.967930   
2909  North Las Vegas  None        1  36.229056 -115.101768   

                    name postal_code  review_count  stars state  
265   Phoenix Valuations       85251             3    3.5    AZ  
1020  Highfield Open Mri       15220             5    2.0    PA  
1332        Ceva Freight       28214             3    1.5    NC  
2399             Poolsaz       85281             4    1.0    AZ  
2909            Amerigas       89030             4    1.0    NV

As you can see, we have the location and name for such businesses, but don't know much else about them. These 368 aren't very useful to us and only represent a small fraction of our data so we'll drop them.

In [29]:
usa_business = usa_business.loc[~usa_business.categories.isnull(), :]

* is the business id unique?

In [30]:
sum(usa_business.business_id.value_counts() > 1)

0

## Save our desired data to file (CSV)

In [31]:
usa_business.to_csv('usa_businesses.csv', index=False)
usa_states_popn.to_csv('usa_population.csv', index=False)

# Summary

In part 1, we've:
* downloaded and uncompressed some data
* read a json file into a pandas DataFrame
* had a very basic look at it
    * its shape/size
    * some sample rows
    * some rows with odd three-character state abbreviations (filtering rows and columns)
* pulled a definitive list of USA states from the web
* and cleaned it
    * the population counts we wanted weren't in a single column
    * the first two rows were not data
* and extracted the desired columns and rows from it
* and then used this to perform some more basic checking of what was present in the data
    * calculating how many USA states we have
    * how many we don't
    * and realised that the data also covers Canada
* having filtered to keep just the businesses in the USA
    * written to file the filtered business data
    * along with our data of USA state populations